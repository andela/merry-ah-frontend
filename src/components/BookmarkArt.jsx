import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { getBookmarkArt } from '../actions/bookmarkActions';

/** Bookmark Component */
class BookmarkArt extends Component {
/**
   * @description fetch bookmark arts when the component
   * is already mounted on the DOM
   * @returns {method} returns the getBookmarkArt method
   */
  componentDidMount() {
    this.props.getBookmarkArt();
  }

  /**
   * @method renderBookmarkArt
   * @description render the bookmark arts
   * @return {JSX} jsx
   */
  renderBookmartArt() {
    if (this.props.bookmark.isFetching) {
      return <p>Loading Bookmarks....</p>;
    }
    if (this.props.bookmark.bookmarks.data.length === 0) {
      return <p>User as 0 bookmark</p>;
    }
    return (
      <div>
        {this.props.bookmark.bookmarks.data.map(bookmark => (
          <div className="container" key={bookmark.id}>
            <div className="row" id="bookmark-list">
              <p className="bookmark-header col-12">Bookmark Arts</p>
              <div className="col-3">
                <div
                  className="my-card"
                  style={{
                    backgroundImage: `url(${bookmark.Art.featuredImg})`,
                  }}
                >
                  <div className="card-category">Poetry</div>
                  <div className="card-container">
                    <div className="card-title">
                      <h6>
                        {' '}
                        <Link to={`/arts/${bookmark.Art.slug}`}>
                          {bookmark.Art.title}
                        </Link>{' '}
                      </h6>
                      <p>{bookmark.Art.description}</p>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        ))}
      </div>
    );
  }

  /**
   * @method render
   * @description renders the bookmark
   * @return {method} return the render bookmark method
   */
  render() {
    return (
      <div className="container">
        <div className="bookmark">{this.renderBookmartArt()}</div>
      </div>
    );
  }
}
BookmarkArt.propTypes = {
  getBookmarkArt: PropTypes.func.isRequired,
  bookmark: PropTypes.shape({
    isFetching: PropTypes.bool.isRequired,
    error: PropTypes.object.isRequired,
    bookmarks: PropTypes.object.isRequired,
  }),
};
const mapStateToProps = state => ({
  bookmark: state.bookmark,
});
export default connect(
  mapStateToProps,
  { getBookmarkArt },
)(BookmarkArt);
