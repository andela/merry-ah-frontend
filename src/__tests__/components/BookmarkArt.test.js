import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import sinon from 'sinon';
import { BookmarkArt } from '../../components/BookmarkArt.jsx';
import { bookmarkProps } from '../__mocks__/propsMock';

let wrapper;
describe('Bookmark Component', () => {
  beforeEach(() => {
    wrapper = mount(
    <MemoryRouter><BookmarkArt {...bookmarkProps}/></MemoryRouter>,
    );
  });
  it('should call renderBookmarkMethod', () => {
    const bookMarkArtInstance = wrapper.find('BookmarkArt');
    sinon.spy(bookMarkArtInstance.instance(), 'renderBookmartArt');
    bookMarkArtInstance.instance().renderBookmartArt();
    expect(bookMarkArtInstance.instance().renderBookmartArt.calledOnce)
      .toEqual(true);
  });
  it('should return loading bookmark when is fetching is true', () => {
    expect(wrapper.find('p').text()).toEqual('Loading Bookmarks....');
  });
  it('should return user as zero bookmark when is fetching is false', () => {
    bookmarkProps.bookmark.isFetching = false;
    bookmarkProps.bookmark.bookmarks.data.length = 0;
    wrapper = mount(<MemoryRouter><BookmarkArt {...bookmarkProps}/></MemoryRouter>);
    expect(wrapper.find('p').text()).toEqual('User has 0 bookmark');
  });
});
