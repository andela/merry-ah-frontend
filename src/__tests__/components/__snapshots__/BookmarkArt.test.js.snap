// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bookmark Component should match snapshot 1`] = `
<MemoryRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "erclkj",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "erclkj",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <BookmarkArt
      bookmark={
        Object {
          "bookmarks": Object {
            "data": Array [
              "addsadd",
              "test",
            ],
          },
          "isFetching": true,
        }
      }
      getBookmarkArt={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
    >
      <div
        className="container"
      >
        <div
          className="bookmark"
        >
          <p>
            Loading Bookmarks....
          </p>
        </div>
      </div>
    </BookmarkArt>
  </Router>
</MemoryRouter>
`;
